Enabled supporting custom truststores for EG dispatch to SSL Enabled Yarn

akshaykjain opened the pull request:
## Related issues/PR:
https://github.com/jupyter/enterprise_gateway/issues/887

## Description:
This PR enables passing the CA bundle path to yarn_api_client Resource manager under the `verify` variable if the said path is exported as an environment variable, like so:
```
export EG_CERT_BUNDLE = "<path_to_custom_CA_Bundle>"
```

If the `EG_CERT_BUNDLE` environment variable is not set, the default behavior is implemented. This ensures backwards compatibility

## Tests
Successfully initiates kernel post making the suggested changes:
_jeg.log_
```
20/09/24 13:52:31 INFO YarnClientImpl: Submitted application application_1598311220321_0081
20/09/24 13:52:31 INFO Client: Application report for application_1598311220321_0081 (state: ACCEPTED)
20/09/24 13:52:31 INFO Client:
         client token: N/A
         diagnostics: [Thu Sep 24 13:52:31 -0700 2020] Application is Activated, waiting for resources to be assigned for AM.  Details : AM Partition = <DEFAULT_PARTITION> ; Partition Resource = <memory:40704, vCores:27> ; Queue's Absolute capacity = 100.0 % ; Queue's Absolute used capacity = 40.74074 % ; Queue's Absolute max capacity = 100.0 % ;
         ApplicationMaster host: N/A
         ApplicationMaster RPC port: -1
         queue: default
         start time: 1600980751030
         final status: UNDEFINED
         tracking URL: http://shad1.fyre.ibm.com:8088/proxy/application_1598311220321_0081/
         user: user1
[D 2020-09-24 13:52:38.098 EnterpriseGatewayApp] Kernel args: {'env': {'PATH': '/opt/ibm/dsxhi/lib/virtualenvs/jeg_base/bin:/usr/lib64/qt-3.3/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/root/bin', 'KERNEL_EXTRA_SPARK_OPTS': '--archives hdfs:///user/dsxhi/environments/83490ad16acca9a76e5b735fd22a37bb8debd9b78a1b9c2c2e9166a9d59bb416/jupyter-py36.tar.gz --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=jupyter-py36.tar.gz/conda/envs/Python-3.6/bin/python3.6  --conf spark.driver.memory=1024m --conf spark.driver.cores=1 --conf spark.executor.instances=1 --conf spark.num.executors=1 --conf spark.executor.memory=1024m --conf spark.executor.cores=1 --conf spark.yarn.appMasterEnv.HI_PYTHONPATH=:jupyter-py36.tar.gz/cc-home/_global_/python-3.6:jupyter-py36.tar.gz/cc-home/_global_/python-3:jupyter-py36.tar.gz/user-home/_global_/python-3.6:jupyter-py36.tar.gz/user-home/_global_/python-3', 'KERNEL_LAUNCH_TIMEOUT': '49', 'KERNEL_USERNAME': 'user1'}, 'kernel_name': 'spark_python_yarn_cluster'}
[I 200924 13:52:38 web:2162] 201 POST /api/kernels?doAs=user1 (127.0.0.1) 15507.34ms
[I 200924 13:52:38 web:2162] 200 GET /api/kernels/502d1669-bba5-4f0b-ab5e-1aaf7b6d68b2?doAs=user1 (127.0.0.1) 2.01ms
```

kevin-bates commented on enterprise_gateway/services/processproxies/yarn.py:
```diff
@@ -64,7 +64,13 @@ def __init__(self, kernel_manager, proxy_config):
+        # export EG_CERT_BUNDLE="/var/lib/cloudera-scm-agent/agent-cert/cm-auto-global_cacerts.pem"
+        cert_bundle = 'EG_CERT_BUNDLE'
+        verify_default = True
+        cert_path = os.getenv(cert_bundle, verify_default)
```
I would prefer we move this to the top of the file, where the other environment variables are defined.  This is a basic convention throughout the repo.
```suggestion
        cert_path = os.getenv('EG_CERT_BUNDLE', True)
```
It might be good to add a comment that `cert_path` defaults to a boolean that can also represent a path.

akshaykjain replied:
sure, addressing this following commit. let me know if that looks good.

kevin-bates reviewed (commented):
Thanks @akshaykjain - just had the one comment about relocating the initialization of `cert_path`.

kevin-bates commented on enterprise_gateway/services/processproxies/yarn.py:
```diff
@@ -24,6 +24,16 @@
+# cert_path: Boolean, defaults to `True`, that controls
+#            whether we verify the server's TLS certificate in yarn-api-client.
+#            Or a string, in which case it must be a path to a CA bundle to use.
+cert_path = os.getenv(cert_bundle, verify_default)
```
Can we please collapse this down to the comments regarding cert_path and its simple initialization?  Users know how to set environment variables and we don't need the intermediate variables (`cert_ bundle` and `verify_default`).  Perhaps including a simple reference that the CA bundle is a pem file would be helpful.
```suggestion
# cert_path: Boolean, defaults to `True`, that controls
#            whether we verify the server's TLS certificate in yarn-api-client.
#            Or a string, in which case it must be a path to a CA bundle (pem file) to use.
cert_path = os.getenv('EG_CERT_BUNDLE', True)
```

akshaykjain replied:
sure

kevin-bates reviewed (commented):
These changes look great and I really appreciate your patience, but I think we should also add an entry to this section of the documentation: https://github.com/jupyter/enterprise_gateway/blob/master/docs/source/config-options.md#addtional-supported-environment-variables

The entries are alphabetical and some are Kubernetes or Container only, and it got me to thinking we should probably change the name from `EG_CERT_BUNDLE` to `EG_YARN_CERT_BUNDLE` since this is only used against YARN and that should probably be explicitly stated in the corresponding paragraph.  I think something close to what is in the code would be sufficient.

I think we can hold off adding anything to the [security features](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/getting-started-security.html#security-features) page for now though.

kevin-bates commented:
Hi Akshay - it is not necessary to squash and force-push your commits.  We will squash commits prior to merge and it confuses the history.  

Also, please include the changes in #890 in this PR - thank you.

akshaykjain commented:
sure no problem. I updated the PR and changed the variable from `EG_CERT_BUNDLE` to `EG_YARN_CERT_BUNDLE` and made the suggested documentation changes.

kevin-bates reviewed (approved):
@akshaykjain - thank you for your patience and great contribution!
